Set -DPy_LIMITED_API flag for py_limited_api=True extensions

janeyx99 opened the pull request:
This could be BC breaking, because there was a period of time when we use py_limited_api=True but don't enforce the flag, and now that we will start enforcing the flag, people's custom extensions may fail to build.

This is strictly still better behavior, as it is sketchy to claim CPython agnosticism without the flag, but calling this out as potential people yelling at us. Ways to mitigate this risk + reasons this may not be too big a deal:
- People haven't known about py_limited_api for extensions much due to lack of docs from python so usage is low right now
- My current tutorial is in store to make new users of py_limited_api pass this flag, so it'd be a noop for them.

Test plan:
* Locally i'm confident as I tried rebuilding ao with this change and it reliably failed (cuz importing torch/extension.h is a nono)
* Unit test wise, the normal python_agnostic one I added should work

______
## BC-breaking note - C++ Extensions `py_limited_api=True` is now built with `-DPy_LIMITED_API`

We formally began respecting the `py_limited_api=True` kwarg in 2.6 and stopped linking libtorch_python.so when the flag was specified, as libtorch_python.so does not guarantee using APIs from from the stable Python limited API. In 2.7, we go further by specifying the `-DPy_LIMITED_API` flag which will enforce that the extension is buildable with the limited API. As a result of this enforcement, **custom extensions that set `py_limited_api=True` but do not abide by the limited API may fail to build**. For an example, see #152243.

This is strictly better behavior as it is sketchy to claim CPython agnosticism without enforcing with the flag. If you run into this issue, please ensure that the extension you are building does not use any APIs which are outside of the Python limited API, e.g., `pybind`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #145784
* __->__ #145764

janeyx99 committed:
Set -DPy_LIMITED_API flag for py_limited_api=True extensions

[ghstack-poisoned]

ezyang reviewed (approved):
test?

albanD commented on torch/utils/cpp_extension.py:
```diff
@@ -573,6 +573,14 @@ def build_extensions(self) -> None:
+                 # Extension constructor with the Py_LIMITED_API flag set to our
+                 # min supported CPython version, here, 3.9.
+                 # See https://docs.python.org/3/c-api/stable.html#c.Py_LIMITED_API
+                self._add_compile_flag(extension, '-DPy_LIMITED_API=0x03090000')
```
I feel like this is the 3rd time I see you hardcoding this constant (0x03090000), let's put it somewhere and re-use :) 
It will make upgrading it easier!

janeyx99 replied:
This is the only usage in our pytorch/pytorch repo tho :p

albanD replied:
It is our lowest supported version, I would expect we do have some checks that people don't run below this?
But yeah I would prefer users of cpp extensions to use our constant value rather than hardcoding theirs. To get the lower version upgrade when we do it.

albanD reviewed (commented):
Sounds good to me!

janeyx99 committed:
Update on "Set -DPy_LIMITED_API flag for py_limited_api=True extensions"

This could be BC breaking, because there was a period of time when we use py_limited_api=True but don't enforce the flag, and now that we will start enforcing the flag, people's custom extensions may fail to build.

This is strictly still better behavior, as it is sketchy to claim CPython agnosticism without the flag, but calling this out as potential people yelling at us. Ways to mitigate this risk + reasons this may not be too big a deal:
- People haven't known about py_limited_api for extensions much due to lack of docs from python so usage is low right now
- My current tutorial is in store to make new users of py_limited_api pass this flag, so it'd be a noop for them.

Test plan:
* Locally i'm confident as I tried rebuilding ao with this change and it reliably failed (cuz importing torch/extension.h is a nono)
* Unit test wise, the normal python_agnostic one I added should work




[ghstack-poisoned]

zou3519 commented on torch/utils/cpp_extension.py:
```diff
@@ -1045,6 +1064,12 @@ def CUDAExtension(name, sources, *args, **kwargs):
         example, to give access to custom ops from python, the library should
         register the ops through the dispatcher.
 
+        Contray to CPython setuptools, who does not define -DPy_LIMITED_API
```
contrary?

albanD reviewed (approved):
Thanks!

janeyx99 committed:
Update on "Set -DPy_LIMITED_API flag for py_limited_api=True extensions"

This could be BC breaking, because there was a period of time when we use py_limited_api=True but don't enforce the flag, and now that we will start enforcing the flag, people's custom extensions may fail to build.

This is strictly still better behavior, as it is sketchy to claim CPython agnosticism without the flag, but calling this out as potential people yelling at us. Ways to mitigate this risk + reasons this may not be too big a deal:
- People haven't known about py_limited_api for extensions much due to lack of docs from python so usage is low right now
- My current tutorial is in store to make new users of py_limited_api pass this flag, so it'd be a noop for them.

Test plan:
* Locally i'm confident as I tried rebuilding ao with this change and it reliably failed (cuz importing torch/extension.h is a nono)
* Unit test wise, the normal python_agnostic one I added should work




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

albanD reviewed (approved):
Ok!
